File: analyzer/src/io/github/colemakmods/keyboard/KeyboardLayout.java
package io.github.colemakmods.keyboard;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents the keyboard layout being analyzed
 *
 * Created by steve on 18/10/14.
 */
public class KeyboardLayout {

    public enum KeyboardType {
        STD, ANGLE, MATRIX_SIMPLE, MATRIX_ERGODOX,
    }

    public final static String ALPHAS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    private String name;
    private List<Key> keyList = new ArrayList<>();
    private int cols;
    private int rows;
    private double[] penaltySameFinger = new double[3];
    private double[][] penaltyNeighbourFinger = new double[4][3];
    private KeyboardType keyboardType = KeyboardType.STD;

    public KeyboardLayout(String name) {
        this.name = name;
    }

    public void addKey(int row, int col, String chars) {
        Key key = new Key(row, col, chars);
        keyList.add(key);
        if (cols < key.getCol() + 1) {
            this.cols = key.getCol() + 1;
        }
        if (rows < key.getRow() + 1) {
            this.rows = key.getRow() + 1;
        }
    }

    public List<String> validate() {
        String alphabet = getAlphabet();
        List <String> errors = new ArrayList<>();
        //check for duplicate characters
        for (char alpha : ALPHAS.toCharArray()) {
            if (alphabet.indexOf(alpha) < 0) {
                errors.add("• Symbol " + alpha + " has no mapping");
            }
        }
        for (int i=0; i < alphabet.length()-1; ++i) {
            for (int j=i+1; j < alphabet.length(); ++j) {
                if (alphabet.charAt(i) == alphabet.charAt(j)) {
                    errors.add("• Symbol " + alphabet.charAt(i) + " has duplicate mapping");
                }
            }
        }
        return errors;
    }

    public Key lookupKey(char c) {
        for (Key key : keyList) {
            if (key.hasChar(c)) {
                return key;
            }
        }
        return null;
    }

    public Key lookupKey(int r, int c) {
        for (Key key : keyList) {
            if (key.getRow() == r && key.getCol() == c) {
                return key;
            }
        }
        return null;
    }

    public String getAlphabet() {
        StringBuilder sb = new StringBuilder();
        for (Key key : keyList) {
            sb.append(key.getChars());
        }
        return sb.toString();
    }

    public String getPrimaryChars() {
        StringBuilder sb = new StringBuilder();
        for (Key key : keyList) {
            sb.append(key.getPrimaryChar());
        }
        return sb.toString();
    }

    public String getName() {
        return name;
    }

    public int getCols() {
        return cols;
    }

    public int getRows() {
        return rows;
    }

    public double getPenaltySameFinger(int rowdiff) {
        return penaltySameFinger[rowdiff];
    }

    public void setPenaltySameFinger(int rowdiff, double penaltyAmount) {
        this.penaltySameFinger[rowdiff] = penaltyAmount;
    }

    public double getPenaltyNeighbourFinger(int outermostFinger, int rowdiff) {
        int outermostFingerLeft = (outermostFinger <= 4) ? outermostFinger : 9 - outermostFinger;
        if (penaltyNeighbourFinger.length <= outermostFingerLeft) return 0;
        return penaltyNeighbourFinger[outermostFingerLeft][rowdiff];
    }

    public void setPenaltyNeighbourFinger(int outermostFinger, int rowdiff, double penaltyAmount) {
        this.penaltyNeighbourFinger[outermostFinger][rowdiff] = penaltyAmount;
    }

    public boolean hasPenaltyNeighbourFinger(int outermostFinger) {
        int outermostFingerLeft = (outermostFinger <= 4) ? outermostFinger : 9 - outermostFinger;
        if (penaltyNeighbourFinger.length <= outermostFingerLeft) return false;
        for (int r = 0; r < 3; ++r) {
            if (penaltyNeighbourFinger[outermostFingerLeft][r] > 0) return true;
        }
        return false;
    }

    public KeyboardLayout.KeyboardType getKeyboardType() {
        return keyboardType;
    }

    public void setKeyboardType(KeyboardLayout.KeyboardType keyboardType) {
        this.keyboardType = keyboardType;
    }

    public void dumpLayout(PrintStream out) {
        out.println();
        out.println("Keyboard '" + name + "' with " + keyList.size() + " keys in " + rows + " rows:");
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                Key key = lookupKey(r, c);
                if (key != null) {
                    out.printf("  %s  ", key.getName());
                } else {
                    out.print("   ");
                }
            }
            out.println();
        }
        out.println("Alphabet " + getAlphabet());
        out.println();
    }

    public void dumpConfig(PrintStream out) {
        out.println("Effort");
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                Key key = lookupKey(r, c);
                if (key != null) {
                    out.printf(" % .1f ", key.getEffort());
                } else {
                    out.print("     ");
                }
            }
            out.println();
        }
        out.println("Fingers");
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                Key key = lookupKey(r, c);
                if (key != null) {
                    out.printf(" %s ", key.getFinger());
                } else {
                    out.print("   ");
                }
            }
            out.println();
        }
        out.println("Same-Finger Penalties");
        for (int rowdiff = 0; rowdiff < 3; ++rowdiff) {
            out.printf(" % .1f ", getPenaltySameFinger(rowdiff));
        }
        out.println();
        out.println("Neighbour-Finger Penalties");
        for (int f = 0; f < 3; ++f) {
            for (int rowdiff = 0; rowdiff < 3; ++rowdiff) {
                out.printf(" % .1f ", getPenaltyNeighbourFinger(f, rowdiff));
            }
            out.println();
        }
        out.println();
    }
}

File: analyzer/src/io/github/colemakmods/keyboard/KeyboardLayout.kt
package io.github.colemakmods.keyboard

import io.github.colemakmods.chars.Alphabet
import java.io.PrintStream

/**
 * Represents the keyboard layout being analyzed
 *
 * Created by steve on 18/10/14.
 */
class KeyboardLayout @JvmOverloads constructor(
    val name: String,
    val keyList: MutableList<Key> = mutableListOf(),
    var cols: Int = 0,
    var rows: Int = 0,
    var penaltySameFinger: DoubleArray = DoubleArray(3),
    var penaltyNeighbourFinger: Array<DoubleArray> = Array(4) { DoubleArray(3) },
    var keyboardType: KeyboardType = KeyboardType.STD,
) {

    enum class KeyboardType {
        STD,
        ANGLE,
        MATRIX_SIMPLE,
        MATRIX_ERGODOX,
    }

    fun duplicate(): KeyboardLayout {
        return KeyboardLayout(
            name, keyList, cols, rows, penaltySameFinger,
            penaltyNeighbourFinger, keyboardType
        )
    }

    fun duplicateWithName(nameName: String): KeyboardLayout {
        return KeyboardLayout(
            nameName, keyList, cols, rows, penaltySameFinger,
            penaltyNeighbourFinger, keyboardType
        )
    }

    fun addKey(row: Int, col: Int, chars: String?) {
        val key = Key(row, col, chars!!)
        keyList.add(key)
        if (cols < key.col + 1) {
            cols = key.col + 1
        }
        if (rows < key.row + 1) {
            rows = key.row + 1
        }
    }

    fun validate(): List<String> {
        val alphabet = generateAlphabet()
        val errors: MutableList<String> = ArrayList()
        errors.addAll(alphabet.findMissingLetters())
        errors.addAll(alphabet.findDuplicateLetters())
        return errors
    }

    fun lookupKey(c: Char): Key? {
        for (key in keyList) {
            if (key.hasChar(c)) {
                return key
            }
        }
        return null
    }

    fun lookupKey(r: Int, c: Int): Key? {
        for (key in keyList) {
            if (key.row == r && key.col == c) {
                return key
            }
        }
        return null
    }

    fun generateAlphabet(): Alphabet {
        val sb = StringBuilder()
        for (key in keyList) {
            sb.append(key.chars)
        }
        return Alphabet(sb.toString())
    }

    fun getPrimaryChars(): String {
        val sb = StringBuilder()
        for (key in keyList) {
            sb.append(key.getPrimaryChar())
        }
        return sb.toString()
    }

    fun getPenaltySameFinger(rowdiff: Int): Double {
        return penaltySameFinger[rowdiff]
    }

    fun setPenaltySameFinger(rowdiff: Int, penaltyAmount: Double) {
        penaltySameFinger[rowdiff] = penaltyAmount
    }

    fun getPenaltyNeighbourFinger(outermostFinger: Int, rowdiff: Int): Double {
        val outermostFingerLeft = if (outermostFinger <= 4) outermostFinger else 9 - outermostFinger
        return if (penaltyNeighbourFinger.size <= outermostFingerLeft) 0.0 else penaltyNeighbourFinger[outermostFingerLeft][rowdiff]
    }

    fun setPenaltyNeighbourFinger(outermostFinger: Int, rowdiff: Int, penaltyAmount: Double) {
        penaltyNeighbourFinger[outermostFinger][rowdiff] = penaltyAmount
    }

    fun hasPenaltyNeighbourFinger(outermostFinger: Int): Boolean {
        val outermostFingerLeft = if (outermostFinger <= 4) outermostFinger else 9 - outermostFinger
        if (penaltyNeighbourFinger.size <= outermostFingerLeft) return false
        for (r in 0..2) {
            if (penaltyNeighbourFinger[outermostFingerLeft][r] > 0) return true
        }
        return false
    }

    fun dumpLayout(out: PrintStream) {
        out.println()
        out.println("Keyboard '" + name + "' with " + keyList.size + " keys in " + rows + " rows:")
        for (r in 0 until rows) {
            for (c in 0 until cols) {
                val key = lookupKey(r, c)
                if (key != null) {
                    out.printf("%s ", key.getName())
                } else {
                    out.print(" ")
                }
            }
            out.println()
        }
        out.println("Alphabet ${generateAlphabet()}")
        out.println()
    }

    fun dumpConfig(out: PrintStream) {
        out.println("Effort")
        for (r in 0 until rows) {
            for (c in 0 until cols) {
                val key = lookupKey(r, c)
                if (key != null) {
                    out.printf(" % .1f ", key.effort)
                } else {
                    out.print("     ")
                }
            }
            out.println()
        }
        out.println("Fingers")
        for (r in 0 until rows) {
            for (c in 0 until cols) {
                val key = lookupKey(r, c)
                if (key != null) {
                    out.printf(" %s ", key.finger)
                } else {
                    out.print("   ")
                }
            }
            out.println()
        }
        out.println("Same-Finger Penalties")
        for (rowdiff in 0..2) {
            out.printf(" % .1f ", getPenaltySameFinger(rowdiff))
        }
        out.println()
        out.println("Neighbour-Finger Penalties")
        for (f in 0..2) {
            for (rowdiff in 0..2) {
                out.printf(" % .1f ", getPenaltyNeighbourFinger(f, rowdiff))
            }
            out.println()
        }
        out.println()
    }

}
